### DISCUSSION TITLE
Database testing

### DISCUSSION THREAD

**Discussion by Anton-Karpenko** [piccolo-orm/piccolo#530] [Q&A] [answered]
Hello, I really enjoy the ORM.

I've got a problem with the database testing. I need to await some database calls in my tests, but it will require making my tests asynchronous. As I understand, piccolo test runner doesn't support that. Is there a way to force to run async queries in a sync way?

I tried to make my test async using fastapi docs for async testing.  Test runner outputs something like:
`RuntimeWarning: coroutine 'TestClass.test_method' was never awaited`

**Comment by dantownsend** [comment, marked as answer]
Hi @Anton-Karpenko, I'm glad you're enjoying Piccolo.

When testing async code, there are a few ways to go about it. If you use [`piccolo tester run`](https://piccolo-orm.readthedocs.io/en/latest/piccolo/projects_and_apps/included_apps.html#tester) it just calls Pytest under the hood.

I see from the [FastAPI docs](https://fastapi.tiangolo.com/advanced/async-tests/) that it recommends using the `pytest.mark.anyio` decorator. Is this what you're currently using?

I tend to use [`pytest-asyncio`](https://pypi.org/project/pytest-asyncio/) instead, as I don't use Trio.

```python
# test_app.py

import pytest

from my_app.tables import MyTable


@pytest.mark.asyncio
async def test_foo():
    rows = await MyTable.select()
    assert len(rows) == 0
``` 

If you prefer using Python's builtin `TestCase`, you can also do things like this:

```python
# test_app.py

from unittest import TestCase

import pytest
from piccolo.utils.sync import run_sync

from my_app.tables import MyTable

class TestFoo(TestCase):
   
    def test_foo(self):
        async def test():
            rows = await MyTable.select()
            self.assertEqual(len(rows), 0)
        
        run_sync(test())
``` 

Does that help at all?

**Reply by Anton-Karpenko** [reply]
Hey, thanks for the response.
I tried to use both variants. The result is the same.
The test raises errors:

```
RuntimeError: Task <Task pending name='Task-42' coro=<TestDB.test_should_create_obj.<locals>.create_obj() running at /app/app/tests/db/test_repositories.py:32> cb=[run_until_complete.<locals>.done_cb()]> got Future <Future pending cb=[Protocol._on_waiter_completed()]> attached to a different loop
asyncpg/protocol/protocol.pyx:168: RuntimeError

During handling of the above exception, another exception occurred:

E   asyncpg.exceptions._base.InterfaceError: cannot perform operation: another operation is in progress

asyncpg/protocol/protocol.pyx:707: InterfaceError
```


```
from app.main import app
from app.db.tables import ObjectModel
from piccolo.utils.sync import run_sync

class TestDB(TestCase):
    def setUp(self):
        create_tables(*TABLES)

    def tearDown(self):
        drop_tables(*TABLES)
        
    def test_should_create_obj():
        data = {...}
        async def create_obj():
            await ObjectModel.insert(ObjectModel(**data))
        run_sync(create_obj())
```

**Reply by dantownsend** [reply]
Are you on v0.74.4 of Piccolo?

**Reply by Anton-Karpenko** [reply]
Yes, I am
`piccolo==0.74.4`

**Reply by dantownsend** [reply]
I'm struggling to recreate it. Do you just have that one simple test, and it still fails?

Do you have any pytest plugins installed?

**Reply by Anton-Karpenko** [reply]
I'll provide a repo with the code, it will be much easier

**Reply by dantownsend** [reply]
Thanks

**Comment by Anton-Karpenko** [comment]
Probably, the issue was in some fixture. I removed all and readded one by one. It led to a situation where my tests were hanging infinity. Here the problem was that I tried to pass an integer value to an FK field and that the related model's id must be UUID instead of int.

Anyway, thanks a lot for the help!

**Reply by dantownsend** [reply]
No worries, glad it's sorted!
